### ISSUE TITLE
Use of node names in XrControllerModelNodePropertiesMSFT is awkward

### ISSUE THREAD

**Issue by Ralith** [KhronosGroup/OpenXR-Docs#89]
`XR_MSFT_controller_model` supplies controller models as GLTF data and identifies GLTF nodes of interest through `XrControllerModelNodePropertiesMSFT`. This identification is made by reference to GLTF node *names*, which is odd:
- GLTF does not require node names to be present or unique, so general-purpose content/tooling may not play well with the use of them as an identifier. For example, [a popular Rust parser](https://github.com/gltf-rs/gltf) doesn't even read names by default.
- GLTF already employs a rigorous scheme for identifying nodes: their index in the nodes array. This is used to e.g. define the node hierarchy.
- Searching for a matching name and parent name is substantially more work than jumping directly to the appropriate node by index.
- The extension dictates a hardcoded name length limit, presumably for simplicity. Further, special meaning is given to an empty `parentNodeName`, even though an empty name is legal in GLTF. Having legal GLTF names be unrepresentable in the extension API adds opportunity for error in authoring controller models. This is particularly concerning in the face of a possible future OpenXR device API, wherein controller models might be supplied by third parties and consumed by diverse runtimes.

@rpavlik mentions that this may have been motivated by existing software discarding node identity information when loading a GLTF file, but preserving names. However, preserving node IDs by a similar mechanism should at least as easy, and it may be advantageous to use node IDs to inform the GLTF decoding directly, e.g. to stash non-animated geometry in a single vertex buffer, and separately allocate each animated component. Finally, mapping node indexes to node (parent) names is much easier than mapping (node parent name, node name) pairs to node indexes, and this can be employed as a last resort by software which for some reason must operate in terms of names.

While this might require changes to existing clients when adopting a successor extension, that should be weighed against additional implementation complexity required from every future application to adopt the extension.

**Comment by bjornbytes** [comment]
I agree with this.  Even though names and indices would both work for me, indices would be preferred.  It's really strange to see glTF nodes identified by their names and their parent names.

**Comment by brycehutchings** [comment]
The requirement of parent names in the MSFT extension was due to legacy reasons and not something we would keep in a future extension. However, I still see the indices as internal "pointers" inside a JSON format and not something that I have seen exposed into an engine. I find it strange that the Rust parser doesn't keep the name fields since they have no purpose besides for accessing from outside the glTF (nothing references by names inside the glTF). From what I remember, glTF 2.0 moved to using indices for performance reasons.

**Comment by Ralith** [comment]
> I find it strange that the Rust parser doesn't keep the name fields since they have no purpose besides for accessing from outside the glTF

I imagine the rationale is that names are non-semantic as far as GLTF is concerned, and hence not required for conformant processing. You can opt in to preserving them, but I've only ever personally used GLTF node names as a diagnostic, if that, so it's easy for me to imagine applications not bothering.

> I still see the indices as internal "pointers" inside a JSON format and not something that I have seen exposed into an engine.

GLTF 2.0 is a relatively young format; it's reasonable to expect these patterns to be a bit in flux still. The controller model extension is metadata inseparable from the associated GLTF data, and it makes sense to view it through a similar lens as a proper GLTF extension. Indeed, an alternative (and basically equivalent) model for an OpenXR extension is to specify a formal GLTF extension which defines an array of node indices, and then allow applications to fetch an array of corresponding poses. I think referring to GLTF data in GLTF terms is the most natural approach, and the most convenient for new software.

As a compromise measure with legacy software that for whatever reason can't be updated to identify GLTF nodes directly, perhaps the future extension could recommend that identified GLTF nodes have globally unique names? Then OpenXR-supplied node indexes can be easily converted into names pre-conversion by applications that have no other choice, while other software can remain simpler/more robust.

**Comment by dgough** [comment]
If people are curious, https://github.com/KhronosGroup/glTF/issues/831 lists the reasons why glTF 2.0 moved to using indices.

glTF 2.0 loader libraries usually expose the indices but not every app is going to be using a glTF 2.0 loader library. Libraries that allow you to load multiple file formats (like obj, fbx or collada) often don't expose the node indices because that is a glTF 2.0 concept. 

For example, Open-Asset-Importer-Lib [aiScene](http://assimp.sourceforge.net/lib_html/structai_scene.html) class exposes the node hierarchy as a tree and I don't think there is a way to get the original node indices if you are loading a glTF 2.0 scene with that library. 

When I looked at different game engines, most imported the non-unique node names but didn't keep or expose the glTF 2.0 node indices.

**Comment by brycehutchings** [comment]
> As a compromise measure with legacy software that for whatever reason can't be updated to identify GLTF nodes directly, perhaps the future extension could recommend that identified GLTF nodes have globally unique names? Then OpenXR-supplied node indexes can be easily converted into names pre-format-conversion by applications that have no other choice, while other software can remain simpler/more robust.

Yes, any future extension would not make the mistakes of the MSFT extension and would require all node names to be unique inside the glTF so they could be identified with a single name and not a (name, parentName) tuple.

If we do see a strong case of indices then we could design the extension to report both and leave it up to the app to use what it wants.

**Comment by kanerogers** [comment]
FWIW, my (admittedly toy) engine [does not read node names](https://github.com/leetvr/hotham/blob/main/hotham/src/gltf_loader.rs) when importing GLTF, as my understanding is that the spec specifically discourages their use.

**Comment by dgough** [comment]
I like Bryce's suggestion of supporting both node indices and unique node names. The runtime would be required to modify the glTF so that it has unique node names but the app would be free to choose to use node indices or node names. Apps that use a glTF loader would probably prefer to use node indices while apps that use a library that can load multiple file formats (like Open-Asset-Importer-Lib) would probably prefer to use unique node names.

**Comment by bjornbytes** [comment]
This should probably be closed now.  `XR_MSFT_controller_model` is deprecated and replaced by `XR_EXT_render_model`, and while the new extension still identifies nodes by name, it guarantees that they are unique and doesn't require looking at parent names.

**Comment by rpavlik** [comment]
Thanks for bringing this up. Yes, this is addressed with the new extension, and we checked with the glTF people to verify that names are how we are meant to refer to nodes. Indices are more of implementation details.
